src/box/box.cc

```diff
@@ -4437,6 +4454,46 @@ box_cfg(void)
 	}
 }
 
+/**
+ * Whether the recovery state triggers for a specific state have already run.
+ */
+static bool recovery_state_reached[RECOVERY_STATE_MAX];
+
+static const char *recovery_state_names[RECOVERY_STATE_MAX] = {
+	/* [RECOVERY_STATE_SNAPSHOT_RECOVERED] = */
+	"snapshot_recovered",
+	/* [RECOVERY_STATE_WAL_RECOVERED] = */
+	"wal_recovered",
+	/* [RECOVERY_STATE_INDEXES_BUILT] = */
+	"indexes_built",
+	/* [RECOVERY_STATE_SYNCED] = */
+	"synced",
+};
+
+int
+box_run_on_recovery_state(enum box_recovery_state state)
+{
+	/* Don't run triggers for the same stage twice. */
```

Comment:
Do I understand correctly that we need this only for the 'sync' state? If so, this looks rather ugly. IMO better add a flag to wherever you invoke the trigger instead.

Reply:
Yep. Done:
<details><summary>diff</summary>

```diff
diff --git a/src/box/box.cc b/src/box/box.cc
index 543b25407..4eafe0d3a 100644
--- a/src/box/box.cc
+++ b/src/box/box.cc
@@ -404,13 +404,18 @@ box_wait_ro(bool ro, double timeout)
        return 0;
 }
 
+/** Whether the triggers for "synced" recovery stage have already run. */
+static bool recovery_state_synced_is_reached;
+
 void
 box_do_set_orphan(bool orphan)
 {
        is_orphan = orphan;
        box_update_ro_summary();
-       if (!is_orphan)
+       if (!is_orphan && !recovery_state_synced_is_reached) {
                box_run_on_recovery_state(RECOVERY_STATE_SYNCED);
+               recovery_state_synced_is_reached = true;
+       }
 }
 
 void
@@ -4398,7 +4403,9 @@ box_cfg_xc(void)
 
        if (is_bootstrap_leader) {
                box_run_on_recovery_state(RECOVERY_STATE_INDEXES_BUILT);
+               assert(!recovery_state_synced_is_reached);
                box_run_on_recovery_state(RECOVERY_STATE_SYNCED);
+               recovery_state_synced_is_reached = true;
        }
        if (!is_bootstrap_leader) {
                replicaset_sync();
@@ -4453,11 +4460,6 @@ box_cfg(void)
        }
 }
 
-/**
- * Whether the recovery state triggers for a specific state have already run.
- */
-static bool recovery_state_reached[RECOVERY_STATE_MAX];
-
 static const char *recovery_state_names[RECOVERY_STATE_MAX] = {
        /* [RECOVERY_STATE_SNAPSHOT_RECOVERED] = */
        "snapshot_recovered",
@@ -4472,10 +4474,6 @@ static const char *recovery_state_names[RECOVERY_STATE_MAX] = {
 int
 box_run_on_recovery_state(enum box_recovery_state state)
 {
-       /* Don't run triggers for the same stage twice. */
-       if (recovery_state_reached[state])
-               return 0;
-       recovery_state_reached[state] = true;
        /*
         * on_recovery_state triggers expect a string, but passing a string
         * literal to them turns out quite ugly, because trigger_run takes
```

</details>